### PR TITLE
fix(ivy): support complex type nodes in ModuleWithProviders

### PR DESCRIPTION
**Pull request by JoostK**
With ngcc's ability to fixup pre-Ivy `ModuleWithProviders` such that they
include a reference to the NgModule type, the type may become a qualified
name:

```
import {ModuleWithProviders} from '@angular/core';
import * as ngcc0 from './module';

export declare provide(): ModuleWithProviders<ngcc0.Module>;
```

ngtsc now takes this situation into account when reflecting a
`ModuleWithProvider`'s type argument.

## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/angular/angular/blob/master/CONTRIBUTING.md#commit
- [x] Tests for the changes have been added (for bug fixes / features)
- [x] Docs have been added / updated (for bug fixes / features)

## PR Type
What kind of change does this PR introduce?

- [x] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Documentation content changes
- [ ] angular.io application / infrastructure changes
- [ ] Other... Please describe:

### CODE REVIEW

File: packages/compiler-cli/test/ngtsc/ngtsc_spec.ts [hunk 1]
```diff
@@ -401,7 +406,8 @@ describe('ngtsc behavioral tests', () => {
     const dtsContents = env.getContents('test.d.ts');
     expect(dtsContents).toContain(`import * as i1 from 'router';`);
```
Comment (JoostK) [inline packages/compiler-cli/test/ngtsc/ngtsc_spec.ts]: Hmm, we have to be careful here. The `InternalRouterModule` is not publicly exported in this test, but still we generated an import to the public API. I think this is correct, as that module is the one that should be exporting that symbol, but it seems to be up to the library author to actually make the type publicly available.

Comment (alxhub) [reply packages/compiler-cli/test/ngtsc/ngtsc_spec.ts]: This is actually something we're aware of. `ngcc` will publicly export these symbols. `ngtsc` will throw an error (which would be a breaking change in Ivy).

**Review by alxhub — APPROVED** [review]

**Comment by petebacondarwin** [comment]
CARETAKER: Can you please run presubmit and then merge.

**Comment by mhevery** [comment]
http://test/OCL:225237261:BASE:225237286:1544648302785:6f9e937

File: packages/compiler-cli/test/ngtsc/ngtsc_spec.ts [hunk 2]
```diff
@@ -387,12 +387,17 @@ describe('ngtsc behavioral tests', () => {
 
     env.write('node_modules/router/index.d.ts', `
         import {ModuleWithProviders} from '@angular/core';
+        import * as internal from './internal';
```
Comment (alxhub) [inline packages/compiler-cli/test/ngtsc/ngtsc_spec.ts]: Let's create a separate test for this case, instead of mutating this one.

**Review by alxhub — CHANGES_REQUESTED** [review]

File: packages/compiler-cli/test/ngtsc/ngtsc_spec.ts (hunk 2 above)
Comment (IgorMinar) [reply packages/compiler-cli/test/ngtsc/ngtsc_spec.ts]: +1 - putting too many assertions into a single test makes it hard to understand the intent of the assertion. please create a separate spec/`it` with clear description. thank you!

**Review by alxhub — APPROVED** [review]

**Comment by petebacondarwin** [comment]
CARETAKER: Can you please run presubmit?

**Pull request closed without merging**